File: typescript/src/parser.js
export class RunnerInfo {
  constructor({ binary, extraArgs = [], thinkingFlags = {}, providerFlag = '', modelFlag = '', agentFlag = '' }) {
    this.binary = binary
    this.extraArgs = extraArgs
    this.thinkingFlags = thinkingFlags
    this.providerFlag = providerFlag
    this.modelFlag = modelFlag
    this.agentFlag = agentFlag
  }
}

export const RUNNER_REGISTRY = {}

function registerDefaults() {
  if (Object.keys(RUNNER_REGISTRY).length) return

  RUNNER_REGISTRY.opencode = new RunnerInfo({
    binary: 'opencode',
    extraArgs: ['run'],
    thinkingFlags: {},
    providerFlag: '',
    modelFlag: '',
    agentFlag: '--agent',
  })
  RUNNER_REGISTRY.claude = new RunnerInfo({
    binary: 'claude',
    extraArgs: [],
    thinkingFlags: {
      0: ['--thinking', 'disabled'],
      1: ['--thinking', 'enabled', '--effort', 'low'],
      2: ['--thinking', 'enabled', '--effort', 'medium'],
      3: ['--thinking', 'enabled', '--effort', 'high'],
      4: ['--thinking', 'enabled', '--effort', 'max'],
    },
    providerFlag: '',
    modelFlag: '--model',
    agentFlag: '--agent',
  })
  RUNNER_REGISTRY.kimi = new RunnerInfo({
    binary: 'kimi',
    extraArgs: [],
    thinkingFlags: {
      0: ['--no-thinking'],
      1: ['--thinking'],
      2: ['--thinking'],
      3: ['--thinking'],
      4: ['--thinking'],
    },
    providerFlag: '',
    modelFlag: '--model',
    agentFlag: '--agent',
  })
  RUNNER_REGISTRY.codex = new RunnerInfo({
    binary: 'codex',
    extraArgs: [],
    thinkingFlags: {},
    providerFlag: '',
    modelFlag: '--model',
  })
  RUNNER_REGISTRY.roocode = new RunnerInfo({
    binary: 'roocode',
    extraArgs: [],
    thinkingFlags: {},
    providerFlag: '',
    modelFlag: '',
  })
  RUNNER_REGISTRY.crush = new RunnerInfo({
    binary: 'crush',
    extraArgs: [],
    thinkingFlags: {},
    providerFlag: '',
    modelFlag: '',
  })

  RUNNER_REGISTRY.oc = RUNNER_REGISTRY.opencode
  RUNNER_REGISTRY.cc = RUNNER_REGISTRY.claude
  RUNNER_REGISTRY.c = RUNNER_REGISTRY.codex
  RUNNER_REGISTRY.cx = RUNNER_REGISTRY.codex
  RUNNER_REGISTRY.k = RUNNER_REGISTRY.kimi
  RUNNER_REGISTRY.rc = RUNNER_REGISTRY.roocode
  RUNNER_REGISTRY.cr = RUNNER_REGISTRY.crush
}

registerDefaults()

const RUNNER_SELECTOR_RE = /^(?:oc|cc|c|cx|k|rc|cr|codex|claude|opencode|kimi|roocode|crush|pi)$/i
const THINKING_RE = /^\+([0-4])$/
const PROVIDER_MODEL_RE = /^:([a-zA-Z0-9_-]+):([a-zA-Z0-9._-]+)$/
const MODEL_RE = /^:([a-zA-Z0-9._-]+)$/
const ALIAS_RE = /^@([a-zA-Z0-9_-]+)$/

export function parseArgs(argv) {
  const parsed = { runner: null, thinking: null, provider: null, model: null, alias: null, prompt: '' }
  const positional = []

  for (const token of argv) {
    if (RUNNER_SELECTOR_RE.test(token) && parsed.runner === null && positional.length === 0) {
      parsed.runner = token.toLowerCase()
    } else if (THINKING_RE.test(token) && positional.length === 0) {
      parsed.thinking = parseInt(token.slice(1), 10)
    } else if (PROVIDER_MODEL_RE.test(token) && positional.length === 0) {
      const m = PROVIDER_MODEL_RE.exec(token)
      parsed.provider = m[1]
      parsed.model = m[2]
    } else if (MODEL_RE.test(token) && positional.length === 0) {
      const m = MODEL_RE.exec(token)
      parsed.model = m[1]
    } else if (ALIAS_RE.test(token) && parsed.alias === null && positional.length === 0) {
      const m = ALIAS_RE.exec(token)
      parsed.alias = m[1]
    } else {
      positional.push(token)
    }
  }

  parsed.prompt = positional.join(' ')
  return parsed
}

function resolveRunnerName(name, config) {
  if (name === null || name === undefined) {
    return config.defaultRunner
  }
  const abbrev = config.abbreviations[name]
  if (abbrev) return abbrev
  return name
}

export function resolveCommand(parsed, config) {
  if (!config) {
    config = {
      defaultRunner: 'oc',
      defaultProvider: '',
      defaultModel: '',
      defaultThinking: null,
      aliases: {},
      abbreviations: {},
    }
  }

  const warnings = []
  let runnerName = resolveRunnerName(parsed.runner, config)
  let info = RUNNER_REGISTRY[runnerName] || RUNNER_REGISTRY[config.defaultRunner] || RUNNER_REGISTRY.opencode

  let aliasDef = null
  if (parsed.alias && config.aliases[parsed.alias]) {
    aliasDef = config.aliases[parsed.alias]
  }

  if (aliasDef && aliasDef.runner && parsed.runner === null) {
    runnerName = resolveRunnerName(aliasDef.runner, config)
    info = RUNNER_REGISTRY[runnerName] || info
  }

  const argv = [info.binary, ...info.extraArgs]

  let effectiveThinking = parsed.thinking
  if (effectiveThinking === null && aliasDef && aliasDef.thinking !== null && aliasDef.thinking !== undefined) {
    effectiveThinking = aliasDef.thinking
  }
  if (effectiveThinking === null || effectiveThinking === undefined) {
    effectiveThinking = config.defaultThinking
  }
  if (effectiveThinking !== null && effectiveThinking !== undefined && effectiveThinking in info.thinkingFlags) {
    argv.push(...info.thinkingFlags[effectiveThinking])
  }

  let effectiveProvider = parsed.provider
  if (effectiveProvider === null && aliasDef && aliasDef.provider) {
    effectiveProvider = aliasDef.provider
  }
  if (effectiveProvider === null) {
    effectiveProvider = config.defaultProvider || null
  }

  let effectiveModel = parsed.model
  if (effectiveModel === null && aliasDef && aliasDef.model) {
    effectiveModel = aliasDef.model
  }
  if (effectiveModel === null) {
    effectiveModel = config.defaultModel || null
  }

  if (effectiveModel && info.modelFlag) {
    argv.push(info.modelFlag, effectiveModel)
  }

  let effectiveAgent = null
  if (parsed.alias && !aliasDef) {
    effectiveAgent = parsed.alias
  }
  if (effectiveAgent === null && aliasDef && aliasDef.agent) {
    effectiveAgent = aliasDef.agent
  }
  if (effectiveAgent) {
    if (info.agentFlag) {
      argv.push(info.agentFlag, effectiveAgent)
    } else {
      warnings.push(
        `warning: runner "${runnerName}" does not support agents; ignoring @${effectiveAgent}`,
      )
    }
  }

  const envOverrides = {}
  if (effectiveProvider) {
    envOverrides.CCC_PROVIDER = effectiveProvider
  }

  const prompt = parsed.prompt.trim()
  if (!prompt) {
    throw new Error('prompt must not be empty')
  }

  argv.push(prompt)
  return { argv, env: envOverrides, warnings }
}

File: go/parser.go
package ccc

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

type RunnerInfo struct {
	Binary        string
	ExtraArgs     []string
	ThinkingFlags map[int][]string
	ProviderFlag  string
	ModelFlag     string
	AgentFlag     string
}

type ParsedArgs struct {
	Runner   *string
	Thinking *int
	Provider *string
	Model    *string
	Alias    *string
	Prompt   string
}

type AliasDef struct {
	Runner   *string
	Thinking *int
	Provider *string
	Model    *string
	Agent    *string
}

type CccConfig struct {
	DefaultRunner   string
	DefaultProvider string
	DefaultModel    string
	DefaultThinking *int
	Aliases         map[string]AliasDef
	Abbreviations   map[string]string
}

var RunnerRegistry map[string]RunnerInfo

var (
	runnerSelectorRE *regexp.Regexp
	thinkingRE       *regexp.Regexp
	providerModelRE  *regexp.Regexp
	modelRE          *regexp.Regexp
	aliasRE          *regexp.Regexp
)

func init() {
	RunnerRegistry = map[string]RunnerInfo{
		"opencode": {
			Binary:    "opencode",
			ExtraArgs: []string{"run"},
			AgentFlag: "--agent",
		},
		"claude": {
			Binary: "claude",
			ThinkingFlags: map[int][]string{
				0: {"--no-thinking"},
				1: {"--thinking", "low"},
				2: {"--thinking", "medium"},
				3: {"--thinking", "high"},
				4: {"--thinking", "max"},
			},
			ModelFlag: "--model",
			AgentFlag: "--agent",
		},
		"kimi": {
			Binary: "kimi",
			ThinkingFlags: map[int][]string{
				0: {"--no-think"},
				1: {"--think", "low"},
				2: {"--think", "medium"},
				3: {"--think", "high"},
				4: {"--think", "max"},
			},
			ModelFlag: "--model",
			AgentFlag: "--agent",
		},
		"codex": {
			Binary:    "codex",
			ModelFlag: "--model",
		},
		"roocode": {
			Binary:    "roocode",
			ModelFlag: "--model",
		},
		"crush": {
			Binary: "crush",
		},
	}
	RunnerRegistry["oc"] = RunnerRegistry["opencode"]
	RunnerRegistry["cc"] = RunnerRegistry["claude"]
	RunnerRegistry["c"] = RunnerRegistry["codex"]
	RunnerRegistry["cx"] = RunnerRegistry["codex"]
	RunnerRegistry["k"] = RunnerRegistry["kimi"]
	RunnerRegistry["rc"] = RunnerRegistry["roocode"]
	RunnerRegistry["cr"] = RunnerRegistry["crush"]

	runnerSelectorRE = regexp.MustCompile(`(?i)^(?:oc|cc|c|cx|k|rc|cr|codex|claude|opencode|kimi|roocode|crush|pi)$`)
	thinkingRE = regexp.MustCompile(`^\+([0-4])$`)
	providerModelRE = regexp.MustCompile(`^:([a-zA-Z0-9_-]+):([a-zA-Z0-9._-]+)$`)
	modelRE = regexp.MustCompile(`^:([a-zA-Z0-9._-]+)$`)
	aliasRE = regexp.MustCompile(`^@([a-zA-Z0-9_-]+)$`)
}

func ptrStr(s string) *string { return &s }
func ptrInt(i int) *int       { return &i }

func ParseArgs(argv []string) ParsedArgs {
	var parsed ParsedArgs
	var positional []string

	for _, token := range argv {
		if runnerSelectorRE.MatchString(token) && parsed.Runner == nil && len(positional) == 0 {
			lower := strings.ToLower(token)
			parsed.Runner = &lower
		} else if m := thinkingRE.FindStringSubmatch(token); m != nil && len(positional) == 0 {
			val, _ := strconv.Atoi(m[1])
			parsed.Thinking = ptrInt(val)
		} else if m := providerModelRE.FindStringSubmatch(token); m != nil && len(positional) == 0 {
			parsed.Provider = &m[1]
			parsed.Model = &m[2]
		} else if m := modelRE.FindStringSubmatch(token); m != nil && len(positional) == 0 {
			parsed.Model = &m[1]
		} else if m := aliasRE.FindStringSubmatch(token); m != nil && parsed.Alias == nil && len(positional) == 0 {
			parsed.Alias = &m[1]
		} else {
			positional = append(positional, token)
		}
	}

	parsed.Prompt = strings.Join(positional, " ")
	return parsed
}

func resolveRunnerName(name *string, config *CccConfig) string {
	if name == nil {
		return config.DefaultRunner
	}
	if abbrev, ok := config.Abbreviations[*name]; ok {
		return abbrev
	}
	return *name
}

func ResolveCommand(parsed ParsedArgs, config *CccConfig) ([]string, map[string]string, []string, error) {
	if config == nil {
		config = DefaultConfig()
	}

	runnerName := resolveRunnerName(parsed.Runner, config)
	info, ok := RunnerRegistry[runnerName]
	if !ok {
		info, ok = RunnerRegistry[config.DefaultRunner]
		if !ok {
			info = RunnerRegistry["opencode"]
		}
	}

	var warnings []string
	var aliasDef *AliasDef
	if parsed.Alias != nil {
		if ad, exists := config.Aliases[*parsed.Alias]; exists {
			aliasDef = &ad
		}
	}

	var requestedAgent *string
	if parsed.Alias != nil && aliasDef == nil {
		requestedAgent = parsed.Alias
	}

	if aliasDef != nil && aliasDef.Runner != nil && parsed.Runner == nil {
		effectiveName := resolveRunnerName(aliasDef.Runner, config)
		if ri, found := RunnerRegistry[effectiveName]; found {
			info = ri
		}
	}

	argv := make([]string, 0, 1+len(info.ExtraArgs))
	argv = append(argv, info.Binary)
	argv = append(argv, info.ExtraArgs...)

	effectiveThinking := parsed.Thinking
	if effectiveThinking == nil && aliasDef != nil && aliasDef.Thinking != nil {
		effectiveThinking = aliasDef.Thinking
	}
	if effectiveThinking == nil {
		effectiveThinking = config.DefaultThinking
	}
	if effectiveThinking != nil {
		if flags, exists := info.ThinkingFlags[*effectiveThinking]; exists {
			argv = append(argv, flags...)
		}
	}

	effectiveProvider := parsed.Provider
	if effectiveProvider == nil && aliasDef != nil && aliasDef.Provider != nil && *aliasDef.Provider != "" {
		effectiveProvider = aliasDef.Provider
	}
	if effectiveProvider == nil && config.DefaultProvider != "" {
		effectiveProvider = ptrStr(config.DefaultProvider)
	}

	effectiveModel := parsed.Model
	if effectiveModel == nil && aliasDef != nil && aliasDef.Model != nil && *aliasDef.Model != "" {
		effectiveModel = aliasDef.Model
	}
	if effectiveModel == nil && config.DefaultModel != "" {
		effectiveModel = ptrStr(config.DefaultModel)
	}

	if effectiveModel != nil && *effectiveModel != "" && info.ModelFlag != "" {
		argv = append(argv, info.ModelFlag, *effectiveModel)
	}

	effectiveAgent := requestedAgent
	if aliasDef != nil && aliasDef.Agent != nil && *aliasDef.Agent != "" {
		effectiveAgent = aliasDef.Agent
	}
	if effectiveAgent != nil && *effectiveAgent != "" {
		if info.AgentFlag != "" {
			argv = append(argv, info.AgentFlag, *effectiveAgent)
		} else {
			warnings = append(warnings, fmt.Sprintf("warning: runner %q does not support agents; ignoring @%s", info.Binary, *effectiveAgent))
		}
	}

	envOverrides := map[string]string{}
	if effectiveProvider != nil && *effectiveProvider != "" {
		envOverrides["CCC_PROVIDER"] = *effectiveProvider
	}

	prompt := strings.TrimSpace(parsed.Prompt)
	if prompt == "" {
		return nil, nil, nil, fmt.Errorf("prompt must not be empty")
	}

	argv = append(argv, prompt)
	return argv, envOverrides, warnings, nil
}
